File: BraintreeCore/src/test/java/com/braintreepayments/api/ConfigurationLoaderUnitTest.java
package com.braintreepayments.api;

import static junit.framework.Assert.assertEquals;
import static org.mockito.ArgumentMatchers.anyInt;
import static org.mockito.ArgumentMatchers.isNull;
import static org.mockito.ArgumentMatchers.same;
import static org.mockito.Matchers.any;
import static org.mockito.Matchers.anyString;
import static org.mockito.Matchers.eq;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import android.util.Base64;

import org.json.JSONException;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.ArgumentCaptor;
import org.powermock.core.classloader.annotations.PowerMockIgnore;
import org.robolectric.RobolectricTestRunner;

@RunWith(RobolectricTestRunner.class)
@PowerMockIgnore({"org.powermock.*", "org.mockito.*", "org.robolectric.*", "android.*", "androidx.*"})
public class ConfigurationLoaderUnitTest {

    private ConfigurationCache configurationCache;

    private BraintreeHttpClient braintreeHttpClient;
    private ConfigurationLoaderCallback callback;

    private Authorization authorization;

    @Before
    public void beforeEach() {
        configurationCache = mock(ConfigurationCache.class);
        authorization = mock(Authorization.class);

        braintreeHttpClient = mock(BraintreeHttpClient.class);
        callback = mock(ConfigurationLoaderCallback.class);
    }

    @Test
    public void loadConfiguration_loadsConfigurationForTheCurrentEnvironment() {
        when(authorization.getConfigUrl()).thenReturn("https://example.com/config");

        ConfigurationLoader sut = new ConfigurationLoader(braintreeHttpClient, configurationCache);
        sut.loadConfiguration(authorization, callback);

        String expectedConfigUrl = "https://example.com/config?configVersion=3";
        ArgumentCaptor<HttpResponseCallback> captor = ArgumentCaptor.forClass(HttpResponseCallback.class);

        verify(braintreeHttpClient).get(eq(expectedConfigUrl), (Configuration) isNull(), same(authorization), eq(HttpClient.RETRY_MAX_3_TIMES), captor.capture());

        HttpResponseCallback httpResponseCallback = captor.getValue();
        httpResponseCallback.onResult(Fixtures.CONFIGURATION_WITH_ACCESS_TOKEN, null);

        verify(callback).onResult(any(ConfigurationLoaderResult.class), (Exception) isNull());
    }

    @Test
    public void loadConfiguration_savesFetchedConfigurationToCache() {
        when(authorization.getConfigUrl()).thenReturn("https://example.com/config");
        when(authorization.getBearer()).thenReturn("bearer");

        ConfigurationLoader sut = new ConfigurationLoader(braintreeHttpClient, configurationCache);
        sut.loadConfiguration(authorization, callback);

        String expectedConfigUrl = "https://example.com/config?configVersion=3";
        ArgumentCaptor<HttpResponseCallback> captor = ArgumentCaptor.forClass(HttpResponseCallback.class);

        verify(braintreeHttpClient).get(eq(expectedConfigUrl), (Configuration) isNull(), same(authorization), eq(HttpClient.RETRY_MAX_3_TIMES), captor.capture());

        HttpResponseCallback httpResponseCallback = captor.getValue();
        httpResponseCallback.onResult(Fixtures.CONFIGURATION_WITH_ACCESS_TOKEN, null);

        String cacheKey = Base64.encodeToString(String.format("%s%s", "https://example.com/config?configVersion=3", "bearer").getBytes(), 0);
        verify(configurationCache).saveConfiguration(any(Configuration.class), eq(cacheKey));
    }

    @Test
    public void loadConfiguration_onJSONParsingError_forwardsExceptionToErrorResponseListener() {
        when(authorization.getConfigUrl()).thenReturn("https://example.com/config");

        ConfigurationLoader sut = new ConfigurationLoader(braintreeHttpClient, configurationCache);
        sut.loadConfiguration(authorization, callback);

        ArgumentCaptor<HttpResponseCallback> captor = ArgumentCaptor.forClass(HttpResponseCallback.class);
        verify(braintreeHttpClient).get(anyString(), (Configuration) isNull(), same(authorization), eq(HttpClient.RETRY_MAX_3_TIMES), captor.capture());

        HttpResponseCallback httpResponseCallback = captor.getValue();
        httpResponseCallback.onResult("not json", null);

        verify(callback).onResult((ConfigurationLoaderResult) isNull(), any(JSONException.class));
    }

    @Test
    public void loadConfiguration_onHttpError_forwardsExceptionToErrorResponseListener() {
        when(authorization.getConfigUrl()).thenReturn("https://example.com/config");

        ConfigurationLoader sut = new ConfigurationLoader(braintreeHttpClient, configurationCache);
        sut.loadConfiguration(authorization, callback);

        ArgumentCaptor<HttpResponseCallback> httpResponseCaptor = ArgumentCaptor.forClass(HttpResponseCallback.class);
        verify(braintreeHttpClient).get(anyString(), (Configuration) isNull(), same(authorization), eq(HttpClient.RETRY_MAX_3_TIMES), httpResponseCaptor.capture());

        HttpResponseCallback httpResponseCallback = httpResponseCaptor.getValue();
        Exception httpError = new Exception("http error");
        httpResponseCallback.onResult(null, httpError);

        ArgumentCaptor<Exception> errorCaptor = ArgumentCaptor.forClass(Exception.class);
        verify(callback).onResult((ConfigurationLoaderResult) isNull(), errorCaptor.capture());

        ConfigurationException error = (ConfigurationException) errorCaptor.getValue();
        assertEquals("Request for configuration has failed: http error",
                error.getMessage());
    }

    @Test
    public void loadConfiguration_whenInvalidToken_forwardsExceptionToCallback() {
        Authorization authorization = new InvalidAuthorization("invalid", "token invalid");

        ConfigurationLoader sut = new ConfigurationLoader(braintreeHttpClient, configurationCache);
        sut.loadConfiguration(authorization, callback);

        ArgumentCaptor<BraintreeException> captor = ArgumentCaptor.forClass(BraintreeException.class);
        verify(callback).onResult((ConfigurationLoaderResult) isNull(), captor.capture());

        BraintreeException exception = captor.getValue();
        assertEquals("token invalid", exception.getMessage());
    }

    @Test
    public void loadConfiguration_whenCachedConfigurationAvailable_loadsConfigurationFromCache() {
        String cacheKey = Base64.encodeToString(String.format("%s%s", "https://example.com/config?configVersion=3", "bearer").getBytes(), 0);

        when(authorization.getConfigUrl()).thenReturn("https://example.com/config");
        when(authorization.getBearer()).thenReturn("bearer");
        when(configurationCache.getConfiguration(cacheKey)).thenReturn(Fixtures.CONFIGURATION_WITH_ACCESS_TOKEN);

        ConfigurationLoader sut = new ConfigurationLoader(braintreeHttpClient, configurationCache);
        sut.loadConfiguration(authorization, callback);

        verify(braintreeHttpClient, times(0)).get(anyString(), (Configuration) isNull(), same(authorization), anyInt(), any(HttpResponseCallback.class));
        verify(callback).onResult(any(ConfigurationLoaderResult.class), (Exception) isNull());
    }
}

File: BraintreeCore/src/test/java/com/braintreepayments/api/ConfigurationLoaderUnitTest.kt
package com.braintreepayments.api

import android.util.Base64
import io.mockk.*
import org.robolectric.RobolectricTestRunner
import org.json.JSONException
import org.junit.Assert.assertEquals
import org.junit.Test
import org.junit.runner.RunWith
import java.lang.Exception

@RunWith(RobolectricTestRunner::class)
class ConfigurationLoaderUnitTest {
    private var configurationCache: ConfigurationCache = mockk(relaxed = true)
    private var braintreeHttpClient: BraintreeHttpClient = mockk(relaxed = true)
    private var callback: ConfigurationLoaderCallback = mockk(relaxed = true)
    private var authorization: Authorization = mockk(relaxed = true)

    @Test
    fun loadConfiguration_loadsConfigurationForTheCurrentEnvironment() {

        every { authorization.configUrl } returns "https://example.com/config"

        val sut = ConfigurationLoader(braintreeHttpClient, configurationCache)
        sut.loadConfiguration(authorization, null, callback)

        val expectedConfigUrl = "https://example.com/config?configVersion=3"
        val callbackSlot = slot<BTHttpResponseCallback>()
        verify {
            braintreeHttpClient.get(
                    expectedConfigUrl,
                    null,
                    authorization,
                    HttpClient.RETRY_MAX_3_TIMES,
                    capture(callbackSlot)
            )
        }

        val httpResponseCallback = callbackSlot.captured
        httpResponseCallback.onResult(BTHttpResponse(body = Fixtures.CONFIGURATION_WITH_ACCESS_TOKEN), null)

        verify { callback.onResult(ofType(Configuration::class), null) }
    }

    @Test
    fun loadConfiguration_savesFetchedConfigurationToCache() {
        every { authorization.configUrl } returns "https://example.com/config"
        every { authorization.bearer } returns "bearer"

        val sut = ConfigurationLoader(braintreeHttpClient, configurationCache)
        sut.loadConfiguration(authorization, null, callback)

        val expectedConfigUrl = "https://example.com/config?configVersion=3"
        val callbackSlot = slot<BTHttpResponseCallback>()
        verify {
            braintreeHttpClient.get(
                    expectedConfigUrl,
                    null,
                    authorization,
                    HttpClient.RETRY_MAX_3_TIMES,
                    capture(callbackSlot)
            )
        }

        val httpResponseCallback = callbackSlot.captured
        httpResponseCallback.onResult(BTHttpResponse(body = Fixtures.CONFIGURATION_WITH_ACCESS_TOKEN), null)
        val cacheKey = Base64.encodeToString(
            "https://example.com/config?configVersion=3bearer".toByteArray(),
            0
        )

        verify {
            configurationCache.saveConfiguration(ofType(Configuration::class), cacheKey)
        }
    }

    @Test
    fun loadConfiguration_onJSONParsingError_forwardsExceptionToErrorResponseListener() {
        every { authorization.configUrl } returns "https://example.com/config"
        val sut = ConfigurationLoader(braintreeHttpClient, configurationCache)
        sut.loadConfiguration(authorization, null, callback)

        val callbackSlot = slot<BTHttpResponseCallback>()
        verify {
            braintreeHttpClient.get(
                    ofType(String::class),
                    null,
                    authorization,
                    HttpClient.RETRY_MAX_3_TIMES,
                    capture(callbackSlot)
            )
        }
        val httpResponseCallback = callbackSlot.captured
        httpResponseCallback.onResult(BTHttpResponse(body = "not json"), null)
        verify {
            callback.onResult(null, ofType(JSONException::class))
        }
    }

    @Test
    fun loadConfiguration_onHttpError_forwardsExceptionToErrorResponseListener() {
        every { authorization.configUrl } returns "https://example.com/config"
        val sut = ConfigurationLoader(braintreeHttpClient, configurationCache)
        sut.loadConfiguration(authorization, null, callback)

        val callbackSlot = slot<BTHttpResponseCallback>()

        verify {
            braintreeHttpClient.get(
                    ofType(String::class),
                    null,
                    authorization,
                    HttpClient.RETRY_MAX_3_TIMES,
                    capture(callbackSlot)
            )
        }

        val httpResponseCallback = callbackSlot.captured
        val httpError = Exception("http error")
        httpResponseCallback.onResult(null, httpError)
        val errorSlot = slot<Exception>()
        verify {
            callback.onResult(null, capture(errorSlot))
        }

        val error = errorSlot.captured as ConfigurationException
        assertEquals(
            "Request for configuration has failed: http error",
            error.message
        )
    }

    @Test
    fun loadConfiguration_whenInvalidToken_forwardsExceptionToCallback() {
        val authorization: Authorization = InvalidAuthorization("invalid", "token invalid")
        val sut = ConfigurationLoader(braintreeHttpClient, configurationCache)
        sut.loadConfiguration(authorization, null, callback)
        val errorSlot = slot<BraintreeException>()
        verify {
            callback.onResult(null, capture(errorSlot))
        }

        val exception = errorSlot.captured
        assertEquals("token invalid", exception.message)
    }

    @Test
    fun loadConfiguration_whenCachedConfigurationAvailable_loadsConfigurationFromCache() {
        val cacheKey = Base64.encodeToString(
            "https://example.com/config?configVersion=3bearer".toByteArray(),
            0
        )
        every { authorization.configUrl } returns "https://example.com/config"
        every { authorization.bearer } returns "bearer"
        every { configurationCache.getConfiguration(cacheKey) } returns Fixtures.CONFIGURATION_WITH_ACCESS_TOKEN

        val sut = ConfigurationLoader(braintreeHttpClient, configurationCache)
        sut.loadConfiguration(authorization, null, callback)

        verify(exactly = 0) {
            braintreeHttpClient.get(
                    ofType(String::class),
                    null,
                    authorization,
                    ofType(Int::class),
                    ofType(BTHttpResponseCallback::class)
            )
        }
        verify { callback.onResult(ofType(Configuration::class), null) }
    }

    @Test
    fun loadConfiguration_withoutCachedConfiguration_sendsAnalyticsEvent() {
        every { authorization.configUrl } returns "https://example.com/config"
        every { authorization.bearer } returns "bearer"

        val apiTiming = mockk<BTAPITiming>()

        every {
            apiTiming.sendEvent(0, 0, "v1/configuration")
        } returns Unit

        val sut = ConfigurationLoader(braintreeHttpClient, configurationCache)
        sut.loadConfiguration(authorization, apiTiming, callback)

        val expectedConfigUrl = "https://example.com/config?configVersion=3"
        val callbackSlot = slot<BTHttpResponseCallback>()
        verify {
            braintreeHttpClient.get(
                    expectedConfigUrl,
                    null,
                    authorization,
                    HttpClient.RETRY_MAX_3_TIMES,
                    capture(callbackSlot)
            )
        }

        val httpResponseCallback = callbackSlot.captured
        httpResponseCallback.onResult(BTHttpResponse(body = Fixtures.CONFIGURATION_WITH_ACCESS_TOKEN), null)

        verify { callback.onResult(ofType(Configuration::class), null) }
    }
}
